ilpy/expressions.py

```diff
@@ -0,0 +1,360 @@
+            try:
+                relation = OPERATOR_MAP[op_type]
+            except KeyError as e:
+                raise ValueError(f"Unsupported comparison operator: {op_type}") from e
+            seen_compare = True
+    return relation
+
+
+def _expression_to_constraint(expr: Expression) -> LinearConstraint:
+    """Convert an expression to a `LinearConstraint`."""
+    constraint = LinearConstraint()
+    if relation := _get_relation(expr):
+        constraint.set_relation(relation)
+    for var, coefficient in _get_coefficients(expr).items():
+        if var is None:
+            # None is the constant term, which is the right hand side of the
+            # comparison.
+            constraint.set_value(-coefficient)
+        elif coefficient != 0:
+            if var.index is None:
+                raise ValueError(
+                    "All variables in a constraint expression must have an index"
+                )
+            constraint.set_coefficient(var.index, coefficient)
+    return constraint
+
+
+def _get_coefficients(
+    expr: Expression | ast.expr,
+    coeffs: dict[Variable | None, float] | None = None,
+    scale: int = 1,
+) -> dict[Variable | None, float]:
+    """Get the coefficients of a linear expression.
+
+    The coefficients are returned as a dictionary mapping Variable to coefficient.
+    The key `None` is used for the constant term.
+
+    Note also that expressions on the right side of a comparison are negated,
+    (so that the comparison is effectively against zero.)
+
+    Args:
+        expr: The expression to get the coefficients of.
+        coeffs: The dictionary to add the coefficients to. If not given, a new
+            dictionary is created.
+        scale: The scale to apply to the coefficients. This is used to negate
+            expressions on the right side of a comparison or scale for multiplication.
+
+    Example:
+
+        >>> u = Variable('u')
+        >>> v = Variable('v')
+        >>> _get_coefficients(2 * u - 5 * v <= 7)
+        {u: 2, v: -5, None: -7}
+
+        coefficients are simplified in the process:
+        >>> _get_coefficients(2 * u - (u + 2 * u) <= 7)
+        {u: -1, None: -7}
+    """
+    if coeffs is None:
+        coeffs = {}
+
+    if isinstance(expr, Compare):
+        if len(expr.ops) != 1:
+            raise ValueError("Only single comparisons are supported")
+        _get_coefficients(expr.left, coeffs)
+        # negate the right hand side of the comparison
+        _get_coefficients(expr.comparators[0], coeffs, -1)
+
+    elif isinstance(expr, BinOp):
+        if isinstance(expr.op, (ast.Mult, ast.Div)):
+            if isinstance(expr.right, Constant):
+                e = expr.left
+                v = expr.right.value
+            elif isinstance(expr.left, Constant):
+                e = expr.right
+                v = expr.left.value
+            else:
+                # XXX: will we ever need multiplication by a variable?
```

Comment:
Yes! For quadratic objectives!

Reply:
yeah, I thought so.  that raises another question:  should the `Expression.constraint()` method be renamed to `linear_constraint()`?  or should it simply be smart and return the right type of constraint given the expression?  (i.e. autodetect whether it's a quadratic expression)

Reply:
also, do you want that done in this PR?  or should I switch it to a `NotImplementedError` for now and follow up in another PR?

Reply:
`NotImplementedError` is totally fine for now.

Reply:
I'd keep the `constraint()` instead of `linear_constraint()`. Future versions might return either, no need for the user to specify it.